proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,138 @@
+---
+simd: '0110'
+title: Exponential fee for write lock accounts
+authors:
+  - Anatoly Yakovenko, Tao Zhu
+category: Standard
+type: Core
+status: Draft
+created: 2024-01-18
+feature: (fill in with feature tracking issues once accepted)
+supersedes: (optional - fill this in if the SIMD supersedes a previous SIMD)
+extends: (optional - fill this in if the SIMD extends the design of a previous
+ SIMD)
+---
+
+## Summary
+
+In a permissionless environment with low fees, users often submit transactions
+for opportunistic trades without considering the success probability. Almost
+half of the Compute Units (CUs) in a block are allocated to failed transactions,
+leading to undesirable scenario where large portion of compute powers primarily
+serving failed DeFi arbitrage transactions. To address this, the proposed
+feature introduces economic back pressure, discouraging spam activities and
+ensuring efficient network resource utilization. It suggests tracking the
+Exponential Moving Average (EMA) of contentious accounts' CU utilization
+per block and exponentially increasing the cost to write-lock these accounts
+if their utilization remains high.
+
+## Motivation
+
+The motivation behind this feature is to introduce economic back pressure,
+dissuading DeFi spammers from overwhelming the network. DeFi spam, defined as
+opportunistic trades with a positive return on investment, currently occupies
+almost half of the CUs in a block as failed transactions. Economic back
+pressure aims to create a deterrent for such spam activities, ensuring network
+resources are efficiently utilized and preventing continuous block congestion
+caused by failed DeFi spam transactions.
+
+## Alternatives Considered
+
+While the priority fee serves to mitigate low-cost spams by decreasing the
+likelihood of less prioritized transactions being included, it cannot entirely
+eliminate the inclusion of spam transactions in a block. As long as there
+remains a chance, no matter how small, to inexpensively include transactions,
```

Comment:
This is a pretty big assumption. I'd like to see the tx scheduler improvements land in 1.18 before this gets seriously considered. Given how the tx scheduler is currently implemented, it's hard to say priority fees aren't sufficient to adjudicate access to blockspace.

Reply:
Am confident improved scheduler in 1.18 will respect priority fee much better, but there is still chance, hight be very small, that lower prio tx lands before higher one.

Reply:
If the probability of low-prio txs getting included is low. Then it seems like the EMA raising fees will mainly affect the non-spam devs/users who will now see raised fees. If non-spam is anywhere close to 6M CUs/block, then non-spam users pay and low-priority spammers will only get hit by high fees very rarely?

Reply:
user will pay more when capacity is reduced; once this is reliably predictive, spammer will have to back off; and normal users pay resources per demand.

Reply:
@crispheaney steady state ingress load on leaders is 100kpps.  The pipeline to dedup/sigverify/fee check, has to handle all that load before it gets to the scheduler.  If it takes more then 400ms, tx isn't getting prioritized in that block.  The only way to get spammers to send fewer txs is to raise the base layer fee.

Reply:
<img width="1425" alt="Screenshot 2024-01-23 at 9 18 31 PM" src="https://github.com/solana-foundation/solana-improvement-documents/assets/1029046/b800914d-4f59-44f8-9c4e-0c2064fbd64f">

Reply:
so if txs take too long to get through the pipeline, prioritization is not effective.  Does that make sense?  Doesn't matter what the scheduler does, if it can't see txs because they are still in the queues.  we need to force sender to stop economically

Reply:
It seems reasonable we focus on improving the throughput scheduling and pre-scheduling stages so that it's more likely txs are able to be prioritized.

If the chance of inclusion for spam is low, then I'm not convinced they will back off.
Fee-payer is can be totally separate from the account funding economic activity for arbitrage.
I can keep my fee-payer balance low and continue to spam. If the fees get too high for my account to fund, then oh well my tx won't make it into the block because the leader won't include me since I can't pay fees.